### PR TITLE
Add math functions overloaded over different numeric types for cuda and hip

### PR DESCRIPTION
**Pull request by bddppq**
@petrex @ashishfarmer @rohithkrn @iotamudelta

### CODE REVIEW

**Comment by bddppq** [comment]
#13521 will add overload for `sincos`

File: c10/cuda/math_compat.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+__MATH_FUNCTIONS_DECL__ float max(float x, float y) {
+  return fmaxf(x, y);
+}
+__MATH_FUNCTIONS_DECL__ double max(double x, double y) {
+  return fmax(x, y);
+}
+
+__MATH_FUNCTIONS_DECL__ float pow(float x, float y) {
```
Comment (dzhulgakov) [inline c10/cuda/math_compat.h]: why no pow with double? :)

File: caffe2/core/common_gpu.h [hunk 2]
```diff
@@ -25,6 +25,8 @@
 #include "caffe2/core/common.h"
 #include "caffe2/core/logging.h"
 
+#include "c10/cuda/math_compat.h"
```
Comment (dzhulgakov) [inline caffe2/core/common_gpu.h]: common_gpu.h is included in a lot of .cc files that are not built with nvcc. Is it fine to have __host__ annotations in them?

**Review by dzhulgakov — APPROVED** [review]

File: caffe2/core/common_gpu.h (hunk 2 above)
Comment (bddppq) [reply caffe2/core/common_gpu.h]: It's actually not good (although it actually works) - gcc doesn't allow it, but cuda headers define both `__host__` and `__device__` to empty macro if `__CUDACC__` is not defined (and I happen to have included my compat file after some cuda includes). I'm going to change it.

**Pull request closed without merging**